src/plugins/liveobjects/livemap.ts

```diff
@@ -174,12 +188,63 @@ export class LiveMap extends LiveObject<LiveMapData> {
     }
 
     this.setRegionalTimeserial(opRegionalTimeserial);
+    this.notifyUpdated(update);
   }
 
   protected _getZeroValueData(): LiveMapData {
     return { data: new Map<string, MapEntry>() };
   }
 
+  protected _updateFromDataDiff(newDataRef: LiveMapData): LiveMapUpdate {
+    const update: LiveMapUpdate = { update: {} };
+
+    for (const [key, currentEntry] of this._dataRef.data.entries()) {
+      // any non-tombstoned properties that exist on a current map, but not in the new data - got deleted
+      if (currentEntry.tombstone === false && !newDataRef.data.has(key)) {
+        update.update[key] = 'deleted';
+      }
+    }
+
+    for (const [key, newEntry] of newDataRef.data.entries()) {
+      if (!this._dataRef.data.has(key)) {
+        // if property does not exist in the current map, but new data has it as a non-tombstoned property - got updated
+        if (newEntry.tombstone === false) {
+          update.update[key] = 'updated';
+          continue;
+        }
+
+        // otherwise, if new data has this prop tombstoned - do nothing, as property didn't exist anyway
+        if (newEntry.tombstone === true) {
+          continue;
+        }
+      }
+
+      // properties that exist both in current and new map data need to have their values compared to decide on the update type
+      const currentEntry = this._dataRef.data.get(key)!;
+
+      // compare tombstones first
+      if (currentEntry.tombstone === true && newEntry.tombstone === false) {
+        // current prop is tombstoned, but new is not. it means prop was updated to a meaningful value
+        update.update[key] = 'updated';
+        continue;
+      }
+      if (currentEntry.tombstone === false && newEntry.tombstone === true) {
+        // current prop is not tombstoned, but new is. it means prop was deleted
+        update.update[key] = 'deleted';
+        continue;
+      }
+
+      // if tombstones are matching, need to compare values with deep equals to see if it was changed
+      const valueChanged = !deepEqual(currentEntry.data, newEntry.data, { strict: true });
```

Comment:
If objects are both tombstones, is that really an `update`?

Reply:
assuming MAP_REMOVE op doesn't send a `data` field for some reason, it is going to `undefined`, just as the local `data`. so `valueChanged = undefined !== undefined; // false`, there will be no update.
however, I can make it as an explicit case just to make sure it always behaves like that

Reply:
changed in https://github.com/ably/ably-js/compare/e0fe4faf6bfba6398581965118e15bba0ba633c7..cf43cb204060344b6606f7c180d41ff5681b32a6